src/dromajo_manycore.cpp

```diff
@@ -0,0 +1,314 @@
+    break;
+    case FIFO_MC_TO_HOST_REQ:
+      fifo_read_success = fifo_read(mc_to_host_req_fifo, fifo_id, val);
+    break;
+    case FIFO_MC_TO_HOST_RESP:
+      fifo_read_success = fifo_read(mc_to_host_resp_fifo, fifo_id, val);
+    break;
+    default:
+    {
+      printf("Undefined/Unknown FIFO type!\nExiting...\n");
+      exit(-1);
+    }
+    break;
+  }
+  return fifo_read_success;
+}
+
+/*
+ * Get the credits for the desired FIFO
+ * Only BP will request the manycore for credits
+ * @params[in] type - Type of FIFO requesting credits
+ * @returns the remaining FIFO credits
+ */
+int mc_fifo_get_credits(mc_fifo_type_t type) {
+  int credits = 0;
+  switch (type) {
+    case FIFO_HOST_TO_MC_REQ:
+      credits = get_fifo_credits(host_to_mc_req_fifo);
+    break;
+    case FIFO_MC_TO_HOST_REQ:
+    case FIFO_MC_TO_HOST_RESP:
+      printf("Manycore will not request credits from the host\n");
+    break;
+    default:
+    {
+      printf("Undefined/Unknown FIFO type!\nExiting...\n");
+      exit(-1);
+    }
+    break;
+  }
+  return credits;
+}
+
+/*
+ * Returns if a given 32-bit or 128-bit FIFO is empty
+ * @params[in] fifo - 128-bit FIFO to check emptiness
+ * @params[in] _empty - Set to true if you need the emptiness of a given 32-bit FIFO
+ * @params[in] fifo_id - Chooses which 32-bit FIFO's emptiness to return
+ * @returns 32-bit/128-bit FIFO emptiness
+ */
+bool get_fifo_empty(mc_fifo_t *fifo, bool _empty, uint32_t fifo_id) {
+  if (_empty)
+    return fifo->fifo[index_map[fifo_id]].empty();
+  else {
+    std::vector<std::queue<uint32_t>>::iterator it;
+    bool is_empty = false;
+    for(it = fifo->fifo.begin(); it != fifo->fifo.end(); ++it) {
+      is_empty |= it->empty();
+    }
+    return is_empty;
+  }
+}
+
+/*
+ * Returns the number of elements in a given 32-bit FIFO
+ * @params[in] fifo - 128-bit FIFO to query size
+ * @params[in] fifo_id - Chooses which 32-bit FIFO's size is queried
+ * @returns number of elements in the FIFO
+ */
+uint32_t get_fifo_size(mc_fifo_t *fifo, uint32_t fifo_id) {
+  return fifo->fifo[index_map[fifo_id]].size();
+}
+
+/*
+ * Returns if a given 32-bit or 128-bit FIFO is full
+ * @params[in] fifo - 128-bit FIFO to check fullness
+ * @params[in] _empty - Set to true if you need the fullness of a given 32-bit FIFO
+ * @params[in] fifo_id - Chooses which 32-bit FIFO's fullness to return
+ * @returns 32-bit/128-bit FIFO fullness
+ */
+bool get_fifo_full(mc_fifo_t *fifo, bool _full, uint32_t fifo_id) {
+  if (_full)
+    return fifo->full[index_map[fifo_id]];
+  else {
+    std::vector<bool>::iterator it;
+    bool is_full = true;
+    for(it = fifo->full.begin(); it != fifo->full.end(); ++it)
+      is_full &= *it;
+    return is_full; 
+  }
+}
+
+/*
+ * Sets the fullness of a given 32-bit FIFO
+ * @params[in] fifo - 128-bit FIFO to set fullness
+ * @params[in] fifo_id - Chooses which 32-bit FIFO's fullness to set
+ */
+void set_fifo_full(mc_fifo_t *fifo, uint32_t fifo_id) {
```

Comment:
Maybe label which methods are intended to be used internally and which are the 'API'?

Reply:
Hopefully the recent commit addresses this.